File: 5_others/3_Maximal_Rectangle.py
"""
LC 85. Maximal Rectangle

Given a rows x cols binary matrix filled with 0's and 1's, find the largest rectangle containing only 1's and return its area.

Example 1:

Input: matrix = [["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]
Output: 6
Explanation: The maximal rectangle is shown in the above picture.
Example 2:

Input: matrix = []
Output: 0
Example 3:

Input: matrix = [["0"]]
Output: 0
Example 4:

Input: matrix = [["1"]]
Output: 1
Example 5:

Input: matrix = [["0","0"]]
Output: 0

Constraints:

rows == matrix.length
cols == matrix.length
0 <= row, cols <= 200
matrix[i][j] is '0' or '1'.
"""
# Time complexity: O(N^2), Space complexity: O(N)
class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return 0
        
        arr = [0] * (len(matrix[0]) + 1)
        maxArea = 0
        
        for r in range(len(matrix)):
            for c in range(len(matrix[0])):
                if matrix[r][c] == "1":
                    arr[c] += 1
                else:
                    arr[c] = 0
                    
            maxArea = self.findMaxRectangle(arr, maxArea)
            
        return maxArea
        
    def findMaxRectangle(self, heights: List[int], area: int) -> int:
        stack = [-1]
        
        for idx in range(len(heights)):
            while heights[idx] < heights[stack[-1]]:
                h = heights[stack.pop()]
                area = max(area, h * (idx - stack[-1] - 1))
            
            stack.append(idx)
        
        return area

File: 1. Problems/7. DP & Hash/3_Maximal_Rectangle.cpp
// Maximal Rectangle

/*
Given a 2D binary matrix filled with 0's and 1's, find the largest rectangle containing only 1's and return its area.

Example:

Input:
[
  ["1","0","1","0","0"],
  ["1","0","1","1","1"],
  ["1","1","1","1","1"],
  ["1","0","0","1","0"]
]
Output: 6
*/

// my solution
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        if (matrix.size() == 0 || matrix[0].size() == 0) return 0;
        
        vector<int> dp ( matrix[0].size(), 0 );
        int max_area = 0;
        for (int i = 0; i < matrix.size(); i++) {
            for (int j = 0; j < matrix[0].size(); j++) {
                if (matrix[i][j] == '1')
                    dp[j]++;
                else
                    dp[j] = 0;
            }
            
            max_area = max(max_area, largestRectangleArea(dp));
        }
        
        return max_area;
    }
    
private:
    int largestRectangleArea(vector<int> &height) {
        stack<int> stk;
        int idx, max_area = 0, area = 0;

        for (idx = 0; idx < height.size();) {
            if (stk.empty() || height[stk.top()] <= height[idx]) {
                stk.push(idx++);
            } else {
                int top = stk.top();
                stk.pop();

                if (stk.empty()) {
                    area = height[top] * idx;
                } else {
                    area = height[top] * (idx - stk.top() - 1);
                }

                max_area = max(max_area, area);
            }
        }

        while (!stk.empty()) {
            int top = stk.top();
            stk.pop();
            
            if (stk.empty()) {
                area = height[top] * idx;
            } else {
                area = height[top] * (idx - stk.top() - 1);
            }

            max_area = max(max_area, area);
        }

        return max_area;
    }
};